src/ServerFactory.php

```diff
@@ -1,151 +0,0 @@
-        SWOOLE_SOCK_UDP,
-        SWOOLE_SOCK_UDP6,
-        SWOOLE_UNIX_DGRAM,
-        SWOOLE_UNIX_STREAM
-    ];
-
-    /**
-     * Enable coroutines within the Swoole HTTP server.
-     *
-     * ONLY available for swoole 4.1.0 or later version.
-     *
-     * When running in coroutine mode, PDO/Mysqli (when Swoole is compiled with
-     * --enable-mysqlnd), Redis, SOAP, file_get_contents, fopen(ONLY TCP, FTP,
-     * HTTP protocol), stream_socket_client, and the fsockopen functions will
-     * automatically switch to a non-blocking, async I/O driver. Avoid blocking
-     * I/O when enabling coroutines.
-     *
-     * @var bool
-     */
-    private $enableCoroutine = false;
-
-    /**
-     * @var string
-     */
-    private $host;
-
-    /**
-     * @var int
-     */
-    private $port;
-
-    /**
-     * @var int
-     */
-    private $mode;
-
-    /**
-     * @var int
-     */
-    private $protocol;
-
-    /**
-     * @var array
-     */
-    private $options = [];
-
-    /**
-     * @var SwooleHttpServer
-     */
-    private $swooleServer;
-
-    /**
-     * @see https://www.swoole.co.uk/docs/modules/swoole-server-methods#swoole_server-__construct
-     * @see https://www.swoole.co.uk/docs/modules/swoole-server/predefined-constants for $mode and $protocol constant
-     * @throws Exception\InvalidArgumentException for invalid $port values
-     * @throws Exception\InvalidArgumentException for invalid $mode values
-     * @throws Exception\InvalidArgumentException for invalid $protocol values
-     */
-    public function __construct(string $host, int $port, int $mode, int $protocol, array $options = [])
-    {
-        if ($port < 1 || $port > 65535) {
-            throw new Exception\InvalidArgumentException('Invalid port');
-        }
-
-        if (! in_array($mode, static::MODES, true)) {
-            throw new Exception\InvalidArgumentException('Invalid server mode');
-        }
-
-        if (! in_array($protocol, static::PROTOCOLS, true)) {
-            throw new Exception\InvalidArgumentException('Invalid server protocol');
-        }
-
-        $this->host = $host;
-        $this->port = $port;
-        $this->mode = $mode;
-        $this->protocol = $protocol;
-        $this->options = $options;
-
-        // If provided, and Swoole 4.1.0 or later is in use, this flag can be
-        // used to enable coroutines for most I/O operations.
-        $this->enableCoroutine = $options['enable_coroutine'] ?? false;
-    }
-
-    /**
-     * Create a swoole server instance
-     *
-     * @see https://www.swoole.co.uk/docs/modules/swoole-server-methods#swoole_server-set for server options
-     */
-    public function createSwooleServer(array $appendOptions = []): SwooleHttpServer
-    {
-        if ($this->swooleServer) {
-            return $this->swooleServer;
-        }
-
-        if ($this->enableCoroutine && method_exists(SwooleRuntime::class, 'enableCoroutine')) {
-            SwooleRuntime::enableCoroutine(true);
```

Comment:
SwooleRuntime::enableCoroutine() is no longer getting called anywhere. Where is this call best placed?
Reading the docs at https://wiki.swoole.com/wiki/page/965.html (And using some Google Translate magic), it's the kind of call you'd put in the entry point of your app (???) - I don't really understand the coroutine stuff, so it's a bit over my head, but as far as the HTTP server is concerned, coroutines are enabled by default inside the server according to the `enable_coroutine` server option here: https://wiki.swoole.com/wiki/page/949.html - I could very well be mistaken though…

Reply:
Different coroutine support entirely. :smile:

The coroutine support in the HTTP server is around the events it triggers and listens to. The `SwooleRuntime::enableCoroutine()` method toggles the value at the runtime engine level, which controls coroutine support for TCP and socket operations (think: calling a remote API from within your code; using TCP operations to talk to a Redis server (as Predis does); etc.).

I'd argue this should go into the HttpServerFactory, as that's our entry point for creating the server instance and governing how it will run; that's also where we have access to the configuration values.

Reply:
Thanks for clearing that up for me! Looking at the deleted ServerFactory, it was getting `enable_coroutine` from the http server options array - which defaults to `true` in the server _(I think)_. So, would it make sense to move the `enable_coroutine` flag up to the top level so that it's effect is unambiguous - i.e. I've set it to `false` instead of deleting the option in one of my test projects, turned off coroutine support in the _server_ and had task workers throw a wobbly.

```php
'zend-expressive-swoole' => [
    'enable_coroutine' => false, // Enable Runtime Coroutine
    'swoole-http-server' => [
        'host' => 'whatever.com',
        'port' => 80,
        'mode' => SWOOLE_PROCESS,
        'options' => [
            // Various HTTP Server Options…
            'enable_coroutine' => true, // Server Coroutine Support
        ],
    ],
],
```

Reply:
Yeah, this would make sense - and is another reason to mark this for a new major version, as users who were enabling the flag in the 1.0 series would need to change the option used.

Reply:
OK. I've just updated the server factory and the docs where it's mentioned